src/common/offline_transformations/src/pruning/propagate_masks.cpp

```diff
@@ -1420,6 +1422,196 @@ class ngraph::pass::mask_propagation::Transpose : public MatcherPass {
+                }
+                total_lengths += split_lengths[i];
+            }
+            if (minus_one_length_idx >= 0 && (*input_mask)[axis].size() > 0) {
+                const auto& input_shape = input.get_partial_shape();
+                if (input_shape[axis].is_dynamic())
+                    return false;
+                auto split_dim = static_cast<uint64_t>(input_shape[axis].get_length());
+                split_lengths[minus_one_length_idx] = split_dim - total_lengths;
+            }
+
+            uint64_t split_start = 0;
+            uint64_t split_end = split_lengths[0];
+            std::vector<Mask::Ptr> output_masks;
+            std::vector<Mask*> output_masks_raw;
+            Mask* input_mask_raw = input_mask.get();
+            for (size_t i = 0; i < split->get_output_size(); i++) {
+                auto output_mask = std::make_shared<Mask>();
+                output_mask->copy_and_slice_mask_from(input_mask_raw, axis, split_start, split_end);
+                output_masks.push_back(output_mask);
+                output_masks_raw.push_back(output_mask.get());
+                output_mask->add_callback(
+                    [input_mask_raw, axis, split_start, split_end](Mask::Ptr cur_mask) -> bool {
+                        cur_mask->copy_and_slice_mask_from(input_mask_raw, axis, split_start, split_end);
+                        return true;
+                    },
+                    input_mask);
+                split_start = split_end;
+                if (i + 1 < split_lengths.size())
+                    split_end += split_lengths[i + 1];
+                setMask(split->output(i), output_masks[i]);
+            }
+
+            auto input_mask_callback = [output_masks_raw, axis, split_lengths](Mask::Ptr cur_mask) -> bool {
+                cur_mask->clean_dim_values();
+                uint64_t split_start = 0;
+                uint64_t split_end = split_lengths[0];
+                for (size_t i = 0; i < output_masks_raw.size(); i++) {
+                    auto mask = output_masks_raw[i];
+                    for (size_t j = 0; j < mask->size(); j++) {
+                        const auto& set = (*mask)[j];
+                        if (j == axis) {
+                            for (auto d : set)
+                                (*cur_mask)[j].insert(d + split_start);
+                        } else if (!set.empty()) {
+                            (*cur_mask)[j] = set;
+                        }
+                    }
+                    split_start = split_end;
+                    if (i + 1 < split_lengths.size())
+                        split_end += split_lengths[i + 1];
+                }
+                cur_mask->initialize_dependencies();
+                return true;
+            };
+
+            for (size_t i = 0; i < split->get_output_size(); i++) {
+                input_mask->add_callback(input_mask_callback, output_masks[i]);
+            }
+            return true;
```

Comment:
You should call callbacks here

Reply:
yes, it looks like we need to call `apply_callback` to propagate masks to subraph above

Reply:
Take a look at the comment here https://github.com/openvinotoolkit/openvino/pull/13506#discussion_r1007440530

Reply:
yes, I've seen it but anyway you need to update masks above (mask1, mask2 and so on) 
![image](https://user-images.githubusercontent.com/21181928/200283519-5d4cba29-1503-4cff-9033-ba17ac3a4c1c.png)
the idea is that there is a tree of masks and if some of them have changed, then we have to update others
for Concat it looks like
![image](https://user-images.githubusercontent.com/21181928/200286739-1476cda1-b426-4fa7-869f-ecf75596a85e.png)
output is connect with only one initialized input mask, the inirialized mask is connected to other input masks, other input masks are connected to the initialized mask but these callbacks always return true

Reply:
Discussed offline, the current approch looks correct